### DISCUSSION TITLE
Structured Logging for Janet (text and JSON)

### DISCUSSION THREAD

**Discussion by dlisboa** [janet-lang/janet#1215] [Show and tell]
Hey everyone. I made an implementation of structured logging for Janet, with support for JSON logs. It follows on some of the ideas of Go's implementation of `slog`. Also stole a little bit from @swlkr `joy` framework.

https://github.com/dlisboa/janet-slog

It was quite fun to make. Let me know if there are any improvements. Contributions welcome!

**Comment by sogaiu** [comment]
Only have some minor remarks (^^;

* I guess ATM buffers and files are supported for logging, but I suppose it might be possible to make it work with things that support a `write` "method" as that is supported by [streams](https://github.com/janet-lang/janet/blob/8183cc5a8dc6fa7beeddb434ed26db966bf8aee5/src/core/net.c#L980) and [other](https://github.com/janet-lang/janet/blob/8183cc5a8dc6fa7beeddb434ed26db966bf8aee5/src/core/net.c#L980) [things](https://github.com/janet-lang/janet/blob/8183cc5a8dc6fa7beeddb434ed26db966bf8aee5/src/core/ev.c#L317)...
* May be `def` would be fine [here](https://github.com/dlisboa/janet-slog/blob/2de7e12fb1660022ee7506dae40f458de0b60a43/test/slog.janet#L13) and [here](https://github.com/dlisboa/janet-slog/blob/2de7e12fb1660022ee7506dae40f458de0b60a43/test/slog.janet#L20)?
* `tuple` is a built-in function and it's shadowed in `format-text` -- may be it doesn't matter

---

Peripheral thing: IIUC, [joy](https://github.com/joy-framework/joy)'s author is actually [swlkr](https://github.com/swlkr/)

**Reply by dlisboa** [reply]
Thanks @sogaiu ! 

Maybe you can help me with the first part:

* That was actually my original idea, but I got confused with how to implement a general interface for them. Files and streams both have `:write`, but files will not flush automatically, so I'd have to call `:flush`, but streams do flush (are unbuffered), and they don't implement `:flush` anyway. Also, `stdout` is a `core/file` but if I `:write` to it it's automatically flushed (`:flush` would just no-op but it's a bit confusing). And buffers support neither, though they're useful mostly for testing I'd still like to use them so there's no cleanup necessary.
* Great, nice catch.
* Indeed, I was aware of it, I just preferred the full descriptive name since the function won't be called there

Sorry @swlkr! Fixed it.

**Reply by sogaiu** [reply]
Not sure of an elegant way to manage things, but may be checking for the existence of `flush` for a "thing" and deciding what to do based on the result might be applied?

```
$ janet
Janet 1.29.1-8183cc5a linux/x64/gcc - '(doc)' for help

repl:1:> (def f (file/temp))
<core/file 0x564FC69B0480>

repl:2:> (get f :flush)
<cfunction file/flush>

repl:3:> (get f :non-method)
nil
```
